### PR TITLE
feat: frontend access controll

### LINKED ISSUE #3049: [BUG] - Any user can Modify group settings
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [X] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

Any user, regardless of their permissions, can access the "Manage Members" page. It makes it pretty hard to, well, manage members as all of them can modify group settings and user permissions (they don't seem to work).

In the logs, when changing the permissions from within a user, a ~~401~~ 403 response is logged. I'd expect the user to not have access to this menu at all.

They can modify the group settings (I don't want users to be able to set the group to public)

### Steps to Reproduce

1) Normal setup
2) Create a group
3) Add users without the Manage, Organize permissions
4) Any user can manage, organize 

### Please provide relevant logs

INFO:     172.18.0.4:38784 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:38794 - "GET /api/admin/groups/eed8cacc-0541-49f0-b58d-884787191799 HTTP/1.1" 200 OK
INFO:     172.18.0.4:38808 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:38816 - "GET /api/admin/groups/eed8cacc-0541-49f0-b58d-884787191799 HTTP/1.1" 200 OK
INFO:     172.18.0.4:38826 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:38842 - "GET /api/users?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46428 - "GET /api/admin/users/03e04860-1d7f-4fb1-b1b2-42901f3f80e1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46438 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46448 - "GET /api/users?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46458 - "GET /api/admin/users/32fb0019-9eb5-4a22-9e17-37d85b6d6d3a HTTP/1.1" 200 OK
INFO:     172.18.0.4:46470 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46486 - "PUT /api/admin/users/32fb0019-9eb5-4a22-9e17-37d85b6d6d3a HTTP/1.1" 200 OK
INFO:     172.18.0.4:46488 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO: 25-Jan-24 22:27:37 	Deleting item with id eed8cacc-0541-49f0-b58d-884787191799
INFO:     172.18.0.4:46496 - "DELETE /api/admin/groups/eed8cacc-0541-49f0-b58d-884787191799 HTTP/1.1" 200 OK
INFO:     172.18.0.4:46512 - "GET /api/admin/groups?page=1&perPage=-1 HTTP/1.1" 200 OK
INFO:     172.18.0.4:56212 - "GET /group/members HTTP/1.1" 307 Temporary Redirect
INFO:     172.18.0.4:56226 - "GET /group/members/ HTTP/1.1" 304 Not Modified
INFO:     172.18.0.4:56240 - "GET /api/users/self HTTP/1.1" 200 OK
INFO:     172.18.0.4:56244 - "GET /api/groups/cookbooks?page=1&perPage=-1&orderBy=position&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:56252 - "GET /api/groups/members HTTP/1.1" 200 OK
INFO:     172.18.0.4:56258 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:56274 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     127.0.0.1:42072 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:58868 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:40356 - "GET /api/admin/about/check HTTP/1.1" 200 OK
INFO:     172.18.0.4:40348 - "GET /api/admin/about HTTP/1.1" 200 OK
INFO:     127.0.0.1:35202 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     127.0.0.1:33546 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     127.0.0.1:60508 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:60146 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     127.0.0.1:47038 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:50556 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:50560 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:50562 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:50596 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:50580 - "GET /api/foods?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:50600 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:50578 - "GET /api/recipes?page=1&perPage=64&orderBy=created_at&orderDirection=desc&paginationSeed=1706218196202&searchSeed=1706218196202&search=&requireAllCategories=false&requireAllTags=false&requireAllTools=false&requireAllFoods=false HTTP/1.1" 200 OK
INFO:     172.18.0.4:37176 - "GET /api/groups/statistics HTTP/1.1" 200 OK
INFO:     172.18.0.4:37186 - "GET /api/users/self/group HTTP/1.1" 200 OK
INFO:     172.18.0.4:37194 - "GET /api/groups/members HTTP/1.1" 200 OK
INFO:     172.18.0.4:37202 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:37206 - "PUT /api/groups/permissions HTTP/1.1" 403 Forbidden
INFO:     172.18.0.4:37208 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:37214 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:37212 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:37224 - "GET /api/recipes?page=1&perPage=64&orderBy=created_at&orderDirection=desc&paginationSeed=1706218206672&searchSeed=1706218206672&search=&requireAllCategories=false&requireAllTags=false&requireAllTools=false&requireAllFoods=false HTTP/1.1" 200 OK
INFO:     172.18.0.4:52686 - "GET /api/users/self/group HTTP/1.1" 200 OK
INFO:     172.18.0.4:52680 - "GET /api/groups/statistics HTTP/1.1" 200 OK
INFO:     172.18.0.4:52698 - "GET /api/groups/members HTTP/1.1" 200 OK
INFO:     172.18.0.4:52702 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52716 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52704 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52728 - "GET /api/recipes?page=1&perPage=64&orderBy=created_at&orderDirection=desc&paginationSeed=1706218210150&searchSeed=1706218210150&search=&requireAllCategories=false&requireAllTags=false&requireAllTools=false&requireAllFoods=false HTTP/1.1" 200 OK
INFO:     172.18.0.4:52742 - "GET /api/groups/statistics HTTP/1.1" 200 OK
INFO:     172.18.0.4:52754 - "GET /api/users/self/group HTTP/1.1" 200 OK
INFO:     172.18.0.4:52798 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52762 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52766 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:52782 - "GET /api/recipes?page=1&perPage=64&orderBy=created_at&orderDirection=desc&paginationSeed=1706218213025&searchSeed=1706218213025&search=&requireAllCategories=false&requireAllTags=false&requireAllTools=false&requireAllFoods=false HTTP/1.1" 200 OK
INFO:     172.18.0.4:52812 - "GET /api/users/self/group HTTP/1.1" 200 OK
INFO:     172.18.0.4:52816 - "GET /api/groups/statistics HTTP/1.1" 200 OK
INFO:     172.18.0.4:52824 - "GET /api/groups/self HTTP/1.1" 200 OK
INFO:     127.0.0.1:36434 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:50188 - "PUT /api/groups/preferences HTTP/1.1" 200 OK
INFO:     172.18.0.4:50198 - "PUT /api/groups/preferences HTTP/1.1" 200 OK
INFO:     172.18.0.4:50204 - "PUT /api/groups/preferences HTTP/1.1" 200 OK
INFO:     127.0.0.1:44310 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     127.0.0.1:59104 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:51708 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:51720 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:51706 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:51696 - "GET /api/recipes?page=1&perPage=64&orderBy=created_at&orderDirection=desc&paginationSeed=1706218304737&searchSeed=1706218304737&search=&requireAllCategories=false&requireAllTags=false&requireAllTools=false&requireAllFoods=false HTTP/1.1" 200 OK
INFO:     172.18.0.4:51734 - "GET /api/users/self/group HTTP/1.1" 200 OK
INFO:     172.18.0.4:51726 - "GET /api/groups/statistics HTTP/1.1" 200 OK
INFO:     127.0.0.1:48550 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.18.0.4:57436 - "POST /api/recipes HTTP/1.1" 201 Created
INFO:     172.18.0.4:57448 - "GET /api/recipes/utyjt HTTP/1.1" 200 OK
INFO:     172.18.0.4:57460 - "GET /api/media/recipes/65185588-bdb5-42a6-9211-e65de7e78909/images/original.webp?rnd=1&version= HTTP/1.1" 404 Not Found
INFO:     172.18.0.4:57480 - "GET /api/recipes/utyjt/comments HTTP/1.1" 200 OK
INFO:     172.18.0.4:57468 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57516 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57504 - "GET /api/units?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57544 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57494 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57556 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57528 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:57542 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:42466 - "PUT /api/recipes/utyjt HTTP/1.1" 200 OK
INFO:     172.18.0.4:42476 - "GET /api/recipes/utyjt/comments HTTP/1.1" 200 OK
INFO:     172.18.0.4:42484 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.18.0.4:42488 - "GET /api/shared/recipes?page=1&perPage=-1&recipe_id=65185588-bdb5-42a6-9211-e65de7e78909 HTTP/1.1" 200 OK
INFO:     172.18.0.4:42504 - "POST /api/shared/recipes HTTP/1.1" 201 Created
INFO:     172.18.0.4:44276 - "GET /g/home/shared/r/3fba6dee-221e-428c-b8af-079dcce1cc60 HTTP/1.1" 200 OK
INFO:     172.18.0.4:44284 - "GET /_nuxt/75f58f5.js HTTP/1.1" 304 Not Modified
INFO:     172.18.0.4:44298 - "GET /api/users/self HTTP/1.1" 200 OK
INFO:     172.18.0.4:44308 - "GET /api/recipes/shared/3fba6dee-221e-428c-b8af-079dcce1cc60 HTTP/1.1" 200 OK
INFO:     172.18.0.4:44322 - "GET /api/recipes/utyjt/comments HTTP/1.1" 200 OK
INFO:     172.18.0.4:44316 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK

### Mealie Version

**Version** v1.0.0
**Build** f0bbe796ce40ef32a4a00eecdfb0efe4eef9da07
**Application Mode** Production 

### Deployment

Docker (Linux)

### Additional Deployment Details

default docker compose file

### LINKED ISSUE #3082: [BUG] - User option "Manage Group Data" does not grant access to "Manage Data" page.
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [X] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

Setting the user flag "User can manage group data" should allow the user access to the manage data pages. The page is currently wrapped with an `AdvancedOnly` so the flag "Enable advanced features" would need to be set for the user to have access to that page.

### Steps to Reproduce

1. sign in as a user which can manage group data 
2. navigate to `/user/profile`
3. no link to manage data page

If you directly visit the page you are able to edit the page even without any flags set. 

### Please provide relevant logs

No relevant logs

### Mealie Version

Version
nightly

Build
[ce110c23e4aff30378624883b5c9e3910ee33736](https://github.com/hay-kot/mealie/commit/ce110c23e4aff30378624883b5c9e3910ee33736)

### Deployment

Docker (Linux)

### Additional Deployment Details

_No response_

### LINKED ISSUE #3083: [BUG] - Any user can change items on manage data page
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [X] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

A user without any of the required flags set can manipulate data in the manage data page when accessing it directly

### Steps to Reproduce

1. Log is as user without any of the user flags set
2. navigate to e.g. `/group/data/foods/` directly
3. you are now able to create, alter & delete any item

### Please provide relevant logs

Not really relevant:

INFO: 30-Jan-24 17:12:17        Deleting item with id cefeeb51-49b9-4ed4-abea-979a8e13feb3
INFO:     192.168.32.2:54574 - "DELETE /api/units/cefeeb51-49b9-4ed4-abea-979a8e13feb3 HTTP/1.1" 200 OK

### Mealie Version

Version
nightly

Build
[ce110c23e4aff30378624883b5c9e3910ee33736](https://github.com/hay-kot/mealie/commit/ce110c23e4aff30378624883b5c9e3910ee33736)

### Deployment

Docker (Linux)

### Additional Deployment Details

_No response_

### LINKED ISSUE #3098: [BUG] - Possible to edit cookbook without being logged in
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [ ] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

Go to https://<yourmealiesite>/cookbooks without being logged in.
There you can edit cookbooks, again without being logged in.
This is a severe security issue.

### Steps to Reproduce

1) Go to https://<yourmealiesite>/cookbooks without being logged in.
2) There you can edit cookbooks, again without being logged in.

### Please provide relevant logs

```
INFO:     127.0.0.1:56788 - "GET /api/app/about HTTP/1.1" 200 OK
INFO:     172.16.40.1:10338 - "GET /api/explore/cookbooks/home?page=1&perPage=-1&orderBy=position&orderDirection=asc HTTP/1.1" 200 OK
INFO:     172.16.40.1:10338 - "GET /sw.js HTTP/1.1" 304 Not Modified
```

### Mealie Version

Version - nightly
Build - 9bf2e3fabd4bb9dce72f472ae13e064b5e9f6b78 

### Deployment

Docker (Linux)

### Additional Deployment Details

_No response_

### PR DESCRIPTION
**Pull request by Kuchenpirat**
## What type of PR is this?

- cleanup
- feature

## What this PR does / why we need it:

This PR adds middlewares to all pages that expect the user to be registered / or have special flags enabled (exept the individual cookbook page)

- [x] Admin (all pages are auth & admin only) already implemented via #2988 
- [ ] g/_groupSlug
	- [x] index.html (already implemented in RecipeExplorerPage)
	- [ ] cookbooks
		- [ ] _slug (depends on wheter cookbook is private or not) which is tricky because if the user does not have  rights to get the cookbook, he will not get anything returned, so its not easy checking wheter he has the rights. 
		- [x] index
	- [x] r
		- [x] _slug
			- [x] index (already implemented)
			- [x] ingredient parser
		- [x] create (incl all sub pages via create.vue)
	- [x] recipes
		- [x] organizers (categories, tags, tools)
		- [x] timeline
	- [x] shared (no restrictions needed)
- [x] group
	- [x] index
	- [x] members
	- [x] migrations
	- [x] notifiers
	- [x] webhooks
	- [x] data
	- [x] mealplan
	- [x] reports (i added only auth here, but realistically nobody would ever guess the right id, and even then the report would not get loaded and there is not much to do here, so it should be fine)
- [x] register (not needed)
- [x] shopping-lists
- [x] user (all pages)
- [x] password-forgot/reset-password/login (no auth needed)

## Which issue(s) this PR fixes:

Fixes #3098
Partially Fixes #3083 (frontend only)
Partially Fixes #3049 (frontend only)
Fixes #3082

## Special notes for your reviewer:

I had to add the middleware to most pages because the middleware will only affect all sub pages if there is a .vue file on the same level with the same name as the folder (see e.g. /group/data). I don't think it would be a huge problem 
This will add frontend restrictions only, we should tighten some backend stuff and check more if the required user flags are set. 

The only page that is not included is the page for individual cookbooks. But this would need a custom integration any way as we need to check whether the cookbook is private or not. 

This does also disable access to the group settings page for any user that does not have the permission to manage group settings. We could also disable all inputs when the user does not have the right permission. Feedback on this is welcome.

I think this can be merged as it, as this is a substantial improvement and a finished product within itself. 

## Testing

Manually tested each page with users that should have access to it, and users that should not (if availlable).

### CODE REVIEW

**Comment by boc-the-git** [comment]
@Kuchenpirat should this still be draft? Looks like some to do items not done as yet.

**Comment by Kuchenpirat** [comment]
> The only page that is not included is the page for individual cookbooks. But this would need a custom integration any way as we need to check whether the cookbook is private or not.

Not really a draft, but have to check out the conflicts once i am back home. (Should be easy tho)
This PR should solve this more broadly for all pages where a general  solution middleware was useable. Which the cookbook page is not one of them.

**Comment by Kuchenpirat** [comment]
should be good to go again👍

File: frontend/middleware/can-manage-only.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+interface CanManageRedirectParams {
+  $auth: any
+  redirect: (path: string) => void
+}
+export default function ({ $auth, redirect }: CanManageRedirectParams) {
+  // If the user is not allowed to organize redirect to the home page
```
Comment (boc-the-git) [inline frontend/middleware/can-manage-only.ts]: Looks like this comment is a copy+paste that isn't applicable to this program

File: frontend/middleware/advanced-only.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+interface AdvancedOnlyRedirectParams {
+    $auth: any
+    redirect: (path: string) => void
+}
+export default function ({ $auth, redirect }: AdvancedOnlyRedirectParams) {
+    // If the user is not allowed to organize redirect to the home page
```
Comment (boc-the-git) [inline frontend/middleware/advanced-only.ts]: Looks like this comment is a copy+paste that isn't applicable to this program

**Comment by Kuchenpirat** [comment]
thanks @boc-the-git. I have updated the comments to reflect what the middleware is actually doing.

**Review by boc-the-git — APPROVED** [review]
Looks good mate, good enhancements 👍

**Pull request merged**